### ISSUE TITLE
`button-name` rule reports a false positive on a button with a non-empty `label`

### ISSUE THREAD

**Issue by professorplumb** [dequelabs/axe-core#4306]
### Product

axe-core

### Product Version

4.8.1,4.8.2

### Latest Version

- [X] I have tested the issue with the latest version of the product

#### Expectation
The following button's label is read correctly by the screen readers I have access to (VoiceOver, NVDA):
```
<label for="the-switch-button">Choose on or off</label>
<button type="button" id="the-switch-button" role="switch" aria-checked="false" />
```

#### Actual
However, running axe-core or axe DevTools against it generates a `button-name` violation.

#### How to Reproduce
1. Run axe DevTools against [this codepen](https://codepen.io/professorplumb/pen/XWGaKpZ).
2. Among the errors from the Codepen page itself will be a "Buttons must have discernible text" error.
3. Start your screen reader and tab to the button. You should here "choose on or off" as the label.
4. (In a full example, the button would be styled as a toggle switch.)

#### Additional context
If this is intended behavior, please advise on the best way to implement this pattern. I'm aware I could duplicate the label text in an `aria-label` attribute on the button, or use `aria-labelledby` to point back at the `label` element, but both seem like jumping through hoops just to placate the tool, when the `label` does the needful for screen readers.

**Comment by straker** [comment]
Thanks for the issue. Labels on buttons have [issues with assistive technologies](https://github.com/dequelabs/axe-core/issues/3696#issuecomment-1665575742), namely with Dragon NaturallySpeaking. Because of this we haven't supported it even though some screen readers do. We've been meaning to add an accessibility warning when we detect labels and buttons, but haven't gotten around to it yet.

Going to close as a duplicate of https://github.com/dequelabs/axe-core/issues/3696

**Comment by professorplumb** [comment]
Thanks for the context and the link @straker!
